src/drawing/text.rs

```diff
@@ -3,37 +3,50 @@ use crate::drawing::Canvas;
 use conv::ValueInto;
 use image::{GenericImage, ImageBuffer, Pixel};
 use std::f32;
-use std::i32;
 
 use crate::pixelops::weighted_sum;
-use rusttype::{point, Font, PositionedGlyph, Rect, Scale};
-use std::cmp::max;
+
+use ab_glyph::{point, Font, GlyphId, OutlinedGlyph, PxScale, Rect, ScaleFont};
 
 fn layout_glyphs(
-    scale: Scale,
-    font: &Font,
+    scale: impl Into<PxScale> + Copy,
+    font: &impl Font,
     text: &str,
-    mut f: impl FnMut(PositionedGlyph, Rect<i32>),
-) -> (i32, i32) {
-    let v_metrics = font.v_metrics(scale);
+    mut f: impl FnMut(OutlinedGlyph, Rect),
+) -> (u32, u32) {
+    let (mut w, mut h) = (0f32, 0f32);
+
+    let font = font.as_scaled(scale);
+    let mut last: Option<GlyphId> = None;
+
+    for c in text.chars() {
+        let glyph_id = font.glyph_id(c);
+        let glyph = glyph_id.with_scale_and_position(scale, point(w, font.ascent()));
+        w += font.h_advance(glyph_id);
+        if let Some(g) = font.outline_glyph(glyph) {
+            if let Some(last) = last {
+                let kern_width = font.kern(glyph_id, last);
+                w += kern_width;
+            }
+            last = Some(glyph_id);
+            let bb = g.px_bounds();
+            let g_height = bb.height();
 
-    let (mut w, mut h) = (0, 0);
+            if g_height > h {
+                h = g_height;
+            }
 
-    for g in font.layout(text, scale, point(0.0, v_metrics.ascent)) {
-        if let Some(bb) = g.pixel_bounding_box() {
-            w = max(w, bb.max.x);
-            h = max(h, bb.max.y);
             f(g, bb);
         }
     }
 
-    (w, h)
+    (w as u32, h as u32)
```

Comment:
Just double-checking that:

- The original `w` and `h` can never be negative?
- The precision loss in converting from floats to ints is acceptable?

Reply:
Actually I am not 100% sure as it would depend on the coordinates and bounds returned by the `ab_glyph` library, but since these are width and height of the text bounds, I'd expect it to be always non-negative, but just for the peace of mind we can keep them as `i32` instead. As for precision loss, I think it should be the same with previous implementation, but since this is a private function, we could return `f32` and carry the precision just a little more.